File: playwright-e2e/pages/TransactionFeedPage.ts
import { expect, type Locator, type Page } from "@playwright/test";

export type FeedName = "everyone" | "friends" | "mine";

const FEED_TEST_IDS: Record<FeedName, string> = {
  everyone: "nav-public-tab",
  friends: "nav-contacts-tab",
  mine: "nav-personal-tab",
};

/** The three transaction feeds behind the tabs on the home screen. */
export class TransactionFeedPage {
  readonly list: Locator;
  readonly skeleton: Locator;
  readonly tabs: Locator;
  readonly newTransaction: Locator;

  constructor(private readonly page: Page) {
    this.list = page.getByTestId("transaction-list");
    this.skeleton = page.getByTestId("list-skeleton");
    this.tabs = page.getByTestId("nav-transaction-tabs");
    this.newTransaction = page.getByTestId("nav-top-new-transaction");
  }

  async goto(): Promise<void> {
    await this.page.goto("/");
    await expect(this.tabs).toBeVisible();
  }

  async openFeed(feed: FeedName): Promise<void> {
    await this.page.getByTestId(FEED_TEST_IDS[feed]).click();
    await this.waitForList();
  }

  /**
   * The list renders a skeleton placeholder while data is in flight, so its
   * disappearance is a deterministic "rendering finished" signal. This is the
   * whole reason the suite contains no `waitForTimeout`.
   */
  async waitForList(): Promise<void> {
    await expect(this.skeleton).toHaveCount(0);
    await expect(this.list).toBeVisible();
  }

  items(): Locator {
    return this.list.locator("li");
  }

  async expectNonEmpty(): Promise<void> {
    await expect(this.items().first()).toBeVisible();
  }

  /**
   * Find the row for a specific transaction and assert its formatted amount.
   *
   * `filter({ hasText })` narrows the list to the matching row and keeps the
   * assertion scoped to it — so a coincidental match elsewhere on the page
   * cannot make this pass by accident.
   */
  async expectTransaction(description: string, formattedAmount: string): Promise<void> {
    const row = this.items().filter({ hasText: description }).first();
    await expect(row).toBeVisible();
    await expect(row).toContainText(formattedAmount);
  }

  async openFirstTransaction(): Promise<void> {
    await this.items().first().click();
    await expect(this.page.getByTestId("transaction-detail-header")).toBeVisible();
  }

  async startNewTransaction(): Promise<void> {
    await this.newTransaction.click();
  }
}

File: selenium-e2e/src/test/java/dev/dsbon/realworld/pages/TransactionFeedPage.java
package dev.dsbon.realworld.pages;

import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

/** The three transaction feeds behind the tabs on the home screen. */
public class TransactionFeedPage extends BasePage {

  private static final By TABS = testId("nav-transaction-tabs");
  private static final By LIST = testId("transaction-list");
  private static final By SKELETON = testId("list-skeleton");
  private static final By ITEMS = By.cssSelector("[data-test='transaction-list'] li");
  private static final By DETAIL_HEADER = testId("transaction-detail-header");
  private static final By NEW_TRANSACTION = testId("nav-top-new-transaction");

  public TransactionFeedPage(WebDriver driver) {
    super(driver);
  }

  public enum Feed {
    EVERYONE("nav-public-tab", "/"),
    FRIENDS("nav-contacts-tab", "/contacts"),
    MINE("nav-personal-tab", "/personal");

    private final String testId;
    private final String path;

    Feed(String testId, String path) {
      this.testId = testId;
      this.path = path;
    }

    By locator() {
      return By.cssSelector("[data-test='%s']".formatted(testId));
    }

    String path() {
      return path;
    }
  }

  public TransactionFeedPage open() {
    navigateTo("/");
    visible(TABS);
    return this;
  }

  /**
   * Switch feeds, and verify the switch actually happened.
   *
   * <p>A plain click here produced the single most misleading failure of this
   * whole suite. The tab click was swallowed by a React re-render, the app stayed
   * on "Everyone", and a payment test then searched the PUBLIC feed for a
   * transaction that was sitting in the personal one. The screenshot showed the
   * money had moved correctly and the balance was right — the test was simply
   * looking in the wrong place, and said "transaction not found".
   *
   * <p>Switching tabs is idempotent, so the looping variant is safe, and the URL
   * is an unambiguous outcome to settle on.
   */
  public TransactionFeedPage openFeed(Feed feed) {
    clickUntil(feed.locator(), d -> feed.path().equals(pathOf(d.getCurrentUrl())));
    awaitList();
    return this;
  }

  private static String pathOf(String url) {
    String path = java.net.URI.create(url).getPath();
    return path == null || path.isEmpty() ? "/" : path;
  }

  /**
   * The list renders a skeleton placeholder while data is in flight, so its
   * disappearance is a deterministic "rendering finished" signal — which is why
   * this suite contains no {@code Thread.sleep}.
   */
  public TransactionFeedPage awaitList() {
    waitUntilAbsent(SKELETON);
    visible(LIST);
    return this;
  }

  public int itemCount() {
    return allOf(ITEMS).size();
  }

  /** The row whose text contains the given description, or empty if absent. */
  public java.util.Optional<WebElement> findByDescription(String description) {
    return wait.until(
        d -> {
          List<WebElement> matches =
              d.findElements(ITEMS).stream().filter(e -> e.getText().contains(description)).toList();
          // Returning empty would end the wait immediately; returning null keeps
          // WebDriverWait polling until the row renders or the timeout expires.
          return matches.isEmpty() ? null : java.util.Optional.of(matches.getFirst());
        });
  }

  /**
   * Open the first row in the current feed.
   *
   * <p>The feed is an infinite list that re-renders as more rows stream in, so a
   * single dispatched click can hit a row React is replacing. Clicking until the
   * detail header appears verifies the real outcome instead of the dispatch.
   */
  public TransactionFeedPage openFirstTransaction() {
    awaitList();
    clickUntil(ITEMS, d -> isVisible(d, DETAIL_HEADER));
    return this;
  }

  public void startNewTransaction() {
    click(NEW_TRANSACTION);
  }
}
